### PR TITLE
Add 'https://' to URLs submitted with no scheme

### PR DESCRIPTION
**Pull request by Xymist**
Plus some linting if that's of any value at all.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -82,9 +82,16 @@ func main() {
 		log.Fatalf(usage)
 	}
 
-	url, err := url.Parse(args[0])
+	uri := args[0]
+	if string(uri[0:3]) != "http" {
```
Comment (davecheney) [inline main.go]: what about strings.HasPrefix ? 

You can also just do uri = "https://" + uri

**Comment by davecheney** [comment]
@Xymist please take a look at #31 as there is probably some overlap between the two pull requests.

**Comment by Xymist** [comment]
MHolt's comments were valuable. I dispute the need for an additional function as mvrilo has used, but you're right. There's a significant overlap.

**Comment by davecheney** [comment]
Fixes #26

File: main.go [hunk 2]
```diff
@@ -82,19 +82,37 @@ func main() {
...
+}
+
 // visit visits a url and times the interaction.
 // If the response is a 30x, visit follows the redirect.
 func visit(url *url.URL) {
 	scheme := url.Scheme
 	hostport := url.Host
+
+	// No hostname, just a port
+	if strings.HasPrefix(hostport, ":") {
```
Comment (davecheney) [inline main.go]: Please move this inside the function at 116. The function at 116 should also be broken out into its own function as it now contains a lot of business logic.

File: main.go [hunk 3]
```diff
@@ -140,6 +162,7 @@ func visit(url *url.URL) {
 			InsecureSkipVerify: insecure,
 		})
 		if err := c.Handshake(); err != nil {
+
```
Comment (davecheney) [inline main.go]: please delete this line

File: main.go [hunk 4]
```diff
@@ -157,7 +180,7 @@ func visit(url *url.URL) {
 		log.Fatalf("unable to create request: %v", err)
 	}
 
-	if err := req.Write(conn); err != nil {
+	if err = req.Write(conn); err != nil {
```
Comment (davecheney) [inline main.go]: please revert this change.

File: main.go [hunk 5]
```diff
@@ -85,38 +85,60 @@ func main() {
...
+	if strings.Contains(uri, "://") != true {
+		if strings.HasSuffix(uri, ":80") != true {
+			return "https://" + uri
+		}
+		return "http://" + uri
+	}
+	return uri
+}
+
+func getHostPort(URLScheme, URLHost string) (string, string, string) {
```
Comment (davecheney) [inline main.go]: Please pass in the callers `*url.URL`

File: main.go (hunk 4 above)
Comment (Xymist) [inline main.go]: err is shadow-declared several times; are you sure?

**Comment by davecheney** [comment]
Positive.

File: main.go [hunk 6]
```diff
@@ -85,38 +85,61 @@ func main() {
...
 
 	visit(url)
 }
 
-// visit visits a url and times the interaction.
-// If the response is a 30x, visit follows the redirect.
-func visit(url *url.URL) {
+func schemify(uri string) string {
+	if strings.Contains(uri, "://") != true {
+		if strings.HasSuffix(uri, ":80") != true {
```
Comment (davecheney) [inline main.go]: Please reverse this logic to make it simpler
```go
if strings.HasSuffix(uri, ":80") {
        return "http://"+uri
}
return "https://"+uri
```

File: main.go [hunk 7]
```diff
@@ -85,38 +85,61 @@ func main() {
 		log.Fatalf(usage)
 	}
 
-	url, err := url.Parse(args[0])
+	uri := schemify(args[0])
```
Comment (davecheney) [inline main.go]: Please break this out into a function
```go
url := parseURL(args[0])
```
Then handle the error condition inside `parseURL`.

File: main.go [hunk 8]
```diff
@@ -85,38 +85,61 @@ func main() {
...
+		log.Fatalf("unsupported url scheme %q", scheme)
+	}
+
+	return scheme, host, port
+}
+
+// visit visits a url and times the interaction.
+// If the response is a 30x, visit follows the redirect.
+func visit(url *url.URL) {
+
```
Comment (davecheney) [inline main.go]: please delete this line

File: main.go [hunk 9]
```diff
@@ -85,38 +85,63 @@ func main() {
...
 	if err != nil {
-		log.Fatalf("could not parse url %q: %v", args[0], err)
+		log.Fatalf("could not parse url %q: %v", uri, err)
 	}
+	return url
+}
 
-	visit(url)
+func schemify(uri string) string {
+	if strings.Contains(uri, "://") != true {
```
Comment (theckman) [inline main.go]: Maybe it's just me, but I find the long-form of this expression unexpected. Any objections to just evaluating the return of the function itself:

```Go
if !strings.Contains(uri, "://") { ... }
```

Comment (Xymist) [inline main.go]: Done.

**Comment by davecheney** [comment]
I'll fix the conflicts and merge this now.
